### ISSUE TITLE
[Question] ESP32 GPIO Control throwing a request timeout

### ISSUE THREAD

**Issue by Sushmitha-rao** [thingsboard/thingsboard#7590]
I am trying to control GPIO pins on ESP32 with GPIO Control widget. My arduino code ias below. The control throws Request timeout and I am not able to communicate with the GPIO.

`#include "DHT.h"
#include <WiFi.h>

#include <ThingsBoard.h>

#define WIFI_AP "FRITZ!Box 7530 @ 433"
#define WIFI_PASSWORD "84546779845604506115"

#define TOKEN "ESP32_TEST"

// DHT
#define DHTPIN 4          
#define DHTTYPE DHT22

#define GPIO21 21
#define GPIO22 22
#define GPIO23 23

#define GPIO21_PIN 42
#define GPIO22_PIN 39
#define GPIO23_PIN 36

char thingsboardServer[] = "192.168.178.52";

WiFiClient wifiClient;

PubSubClient client(wifiClient);
// Initialize DHT sensor.
DHT dht(DHTPIN, DHTTYPE);

ThingsBoard tb(wifiClient);

int status = WL_IDLE_STATUS;
unsigned long lastSend;

// We assume that all GPIOs are LOW
boolean gpioState[] = {false, false, false};

void setup()
{
  Serial.begin(115200);
  dht.begin();

  pinMode(GPIO21, OUTPUT);
  pinMode(GPIO22, OUTPUT);
  pinMode(GPIO23, OUTPUT);
  
  delay(10);
  InitWiFi();
  lastSend = 0;

  client.setServer( thingsboardServer, 1883);
  client.setCallback(on_message);
}

void loop()
{
  if ( !client.connected() ) {
    reconnect();
  }

  client.loop();

  /*if ( !tb.connected() ) {
    reconnect();
  }

  if ( millis() - lastSend > 1000 ) { // Update and send only after 1 seconds
    getAndSendTemperatureAndHumidityData();
    lastSend = millis();
  }

  tb.loop();*/
}

void getAndSendTemperatureAndHumidityData()
{
  Serial.println("Collecting temperature data.");

  // Reading temperature or humidity takes about 250 milliseconds!
  float humidity = dht.readHumidity();
  // Read temperature as Celsius (the default)
  float temperature = dht.readTemperature();
  float gasSensorValue = analogRead(35);

  // Check if any reads failed and exit early (to try again).
  if (isnan(humidity) || isnan(temperature)) {
    Serial.println("Failed to read from DHT sensor!");
    return;
  }

  Serial.println("Sending data to ThingsBoard:");
  Serial.print("Humidity: ");
  Serial.print(humidity);
  Serial.print(" %\t");
  Serial.print("Temperature: ");
  Serial.print(temperature);
  Serial.println(" *C ");

  Serial.println("Gas Sensor Value: ");
  Serial.print(gasSensorValue);

  tb.sendTelemetryFloat("LPG Concentration in ppm", gasSensorValue);
  tb.sendTelemetryFloat("Temperature in C", temperature);
  tb.sendTelemetryFloat("Humidity in %", humidity);
}

void on_message(const char* topic, byte* payload, unsigned int length) {

  Serial.print("On message");

  char json[length + 1];
  strncpy (json, (char*)payload, length);
  json[length] = '\0';

  Serial.print("Topic: ");
  Serial.println(topic);
  Serial.print("Message: ");
  Serial.println(json);

  // Decode JSON request
  StaticJsonDocument<200> jsonDoc;
  DeserializationError err = deserializeJson(jsonDoc, (char*)json);

  if (err)
  {
    Serial.print("Deserialization() failed with code:");
    Serial.println(err.c_str());
    return;
  }

  // Check request method
  String methodName = String((const char*)jsonDoc["method"]);

  if (methodName.equals("getGpioStatus")) {

    Serial.print("Get Method");
    // Reply with GPIO status
    String responseTopic = String(topic);
    responseTopic.replace("request", "response");
    client.publish(responseTopic.c_str(), get_gpio_status().c_str());
  } else if (methodName.equals("setGpioStatus")) {
    // Update GPIO status and reply
    set_gpio_status(jsonDoc["params"]["pin"], jsonDoc["params"]["enabled"]);
    String responseTopic = String(topic);
    responseTopic.replace("request", "response");
    client.publish(responseTopic.c_str(), get_gpio_status().c_str());
    client.publish("v1/devices/me/attributes", get_gpio_status().c_str());
  }
}

String get_gpio_status() {
  Serial.print("Inside get");
  // Prepare gpios JSON payload string
  StaticJsonDocument<200> jsonDoc;

  JsonObject data = jsonDoc.to<JsonObject>();

  data[String(GPIO21_PIN)] = gpioState[0] ? true : false;
  data[String(GPIO22_PIN)] = gpioState[1] ? true : false;
  data[String(GPIO23_PIN)] = gpioState[2] ? true : false;
  char payload[256];
  serializeJson(jsonDoc, payload);
  //data.printTo(payload, sizeof(payload));
  String strPayload = String(payload);
  Serial.print("Get gpio status: ");
  Serial.println(strPayload);
  return strPayload;
}

void set_gpio_status(int pin, boolean enabled) {
  Serial.print("Inside Set");
  if (pin == GPIO21_PIN) {
    // Output GPIOs state
    digitalWrite(GPIO21, enabled ? HIGH : LOW);
    // Update GPIOs state
    gpioState[0] = enabled;
  } else if (pin == GPIO22_PIN) {
    // Output GPIOs state
    digitalWrite(GPIO22, enabled ? HIGH : LOW);
    // Update GPIOs state
    gpioState[1] = enabled;
  } else if (pin == GPIO23_PIN) {
    // Output GPIOs state
    digitalWrite(GPIO23, enabled ? HIGH : LOW);
    // Update GPIOs state
    gpioState[2] = enabled;
  }
}

void InitWiFi()
{
  Serial.println("Connecting to AP ...");
  // attempt to connect to WiFi network

  WiFi.mode(WIFI_STA);
  WiFi.begin(WIFI_AP, WIFI_PASSWORD);
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("Connected to AP");
}

void reconnect() {
  // Loop until we're reconnected
  while (!tb.connected()) {
    status = WiFi.status();
    if ( status != WL_CONNECTED) {
      WiFi.begin(WIFI_AP, WIFI_PASSWORD);
      while (WiFi.status() != WL_CONNECTED) {
        delay(500);
        Serial.print(".");
      }
      Serial.println("Connected to AP");
    }
    Serial.print("Connecting to ThingsBoard node ...");
    if ( tb.connect(thingsboardServer, TOKEN) ) {
      Serial.println( "[DONE]" );
      client.subscribe("v1/devices/me/rpc/request/+");
      // Sending current GPIO status
      Serial.println("Sending current GPIO status ...");
      client.publish("v1/devices/me/attributes", get_gpio_status().c_str());
    } else {
      Serial.print( "[FAILED]" );
      Serial.println( " : retrying in 5 seconds]" );
      // Wait 5 seconds before retrying
      delay( 5000 );
    }
  }
}
```

**Comment by mykhailokornieiev** [comment]
As the community has not provided assistance on this matter over an extended period, I am closing this ticket due to inactivity. If it is still relevant, feel free to open a new ticket with a full description of your request.
